### DISCUSSION TITLE
How to add buld device bays

### DISCUSSION THREAD

**Discussion by swash110** [netbox-community/netbox#15062] [Help Wanted!]
I am very new to the DCIM tool and not much familiar with coding :) I have created a device and now trying to add bulk device bays using CSV file , I also tried by converting CSV to JSON but everytime I get the error which says "Device bay has no device" . I did create the device first but not sure what I am missing. 
![Device created](https://github.com/netbox-community/netbox/assets/159192194/64a4662a-6465-49d0-96b0-76aadbe67592)
![Error message](https://github.com/netbox-community/netbox/assets/159192194/9591f9b0-b0f7-4e62-954d-d0fcd222d2fe)

**Comment by candlerb** [comment]
If you don't show your CSV data, then it's hard for us to see what's wrong with it (or whether you have come across a bug).

However, I can get the same error as you (in v3.7.2) by uploading a device bay without specifying the device it belongs to:

```
name
foo
```

Or by giving the name of a non-existent device:

```
name,device
foo,blah-does-not-exist
```

Hence it looks like your uploaded data is invalid, but I'd also say it is a bug in Netbox that it gives an exception instead of a proper error message. Raised as #15067

**Reply by swash110** [reply]
Hi , Attached here is my CSV file.
[AMHRNSDWP1 Shelf 1.csv](https://github.com/netbox-community/netbox/files/14195610/AMHRNSDWP1.Shelf.1.csv)

**Reply by candlerb** [reply]
The column headings have the wrong capitalization: they should be "name" not "Name" etc.  The correct capitalization is shown under the heading "Field options" at `/dcim/device-bays/import/`. (Note: CSV _export_ uses different capitalization, and in many other ways is not compatible with CSV import. Please don't argue - that's just how it is. It's been complained about many times before)

From your original screenshot, is the site name "AMHRNSP1DW" and the device name is "AMHRNSP1DW (Shelf 1)"? If so, the rest looks OK to me, and upload should work.

But there are two other structural problems I see.

One is that you should not be using Device Bays to model things like fans and cards; you should be using Modules in Module Bays.

Device Bays are for completely independent Devices within a housing, which have their own management IP addresses, like blade servers within a blade chassis.

Items within a device, such as fans, PSUs, and line cards, are either Modules or Inventory Items. (The difference is that Modules can have their own components such as interfaces and console ports and go into named bays, whilst inventory items are normally used for simpler things like DIMMs and hard drives)

Secondly, the things you are uploading like "100G MOTR WL3n Basic C-Band 10xSFP+" should be Modules, not Module Bays.  The Module Bays are the slots where Modules go.  For example, the device should have a Module Bay called "1-10", and then you should upload a Module into module bay "1-10" of Module Type "100G MOTR WL3n Basic C-Band 10xSFP+" with name "100g-1" (or whatever).

In fact, to do this properly, the Device Type would have Module Bay Templates for each of the slots. Then when you create an instance of the Device, it automatically gets all the right Module Bays. *Then* you upload actual Module instances to go into those bays.

The Module Type "100G MOTR WL3n Basic C-Band 10xSFP+" would also likely have ten interface templates, with names like `TenGig{module}/1`, `TenGig{module}/2` etc. This would automatically create all the interfaces for you to connect to, when the module is instantiated.

But I've wandered very far off the scope of your original issue, so I'm going to leave it at that.

**Reply by swash110** [reply]
Thanks a lot for the quick reply and appreciate your comment on modeling structure with details. I will proceed with your recommendations as its making more sense to me.
